modal/stub.py

```diff
@@ -367,36 +367,27 @@ async def serve(self, client=None, stdout=None, show_progress=None, timeout=None
             except asyncio.exceptions.CancelledError:
                 return
         else:
-            event_agen = watch(self._local_mounts, output_mgr, timeout)
-            event = await event_agen.__anext__()
-
-            curr_proc = None
-            try:
-                async with self._run(client, output_mgr, None, mode=StubRunMode.SERVE) as app:
-                    client.set_pre_stop(app.disconnect)
-                    existing_app_id = app.app_id
-                    event = await event_agen.__anext__()
-                    if sys.version_info.major == 3 and sys.version_info.minor <= 7:
-                        while event != AppChange.TIMEOUT:
+            async with self._run(client, output_mgr, None, mode=StubRunMode.SERVE) as app:
+                client.set_pre_stop(app.disconnect)
+                existing_app_id = app.app_id
+
+                curr_proc = None
+                try:
+                    async for _ in watch(self._local_mounts, output_mgr, timeout):
+                        if sys.version_info.major == 3 and sys.version_info.minor <= 7:
                             output_mgr.print_if_visible(
                                 "Live-reload skipped. This feature is unsupported below Python 3.8. Upgrade to Python 3.8+ to enable live-reloading."
                             )
-                            event = await event_agen.__anext__()
-                        return
-
-                # live-reloading loop
-                while event != AppChange.TIMEOUT:
-                    curr_proc = await restart_serve(
-                        existing_app_id=app.app_id, prev_proc=curr_proc, output_mgr=output_mgr
-                    )
-                    event = await event_agen.__anext__()
-            finally:
-                if curr_proc:
-                    try:
-                        curr_proc.send_signal(signal.SIGINT)
-                    except ProcessLookupError:
-                        logger.warning("Could not interrupt app serve. Supervised process already terminated.")
-                await event_agen.aclose()
+                        else:
+                            curr_proc = await restart_serve(
```

Comment:
Because the restart happens inside the context manager, you'll get double logging here I think. the `__exit__` on the context manager closes the logs loop before the new one starts up. 

It's an complexity in the code that worth me commenting, in retrospect.